### ISSUE TITLE
reindex timeout

### ISSUE THREAD

**Issue by mickeyn** [elastic/elasticsearch-perl#112]
Hi,
when cloning a large index using the reindex method (2.03):
```perl
            my $bulk = $self->es->bulk_helper(
                index => $dst_idx,
                type  => $type,
            );
            $bulk->reindex( source => { index => $self->index->name }, );
            $bulk->flush; # needed???
```

I got a timeout failure:
```
[Timeout] ** [http://127.0.0.1:9200]-[599] Timed out while waiting for socket to become ready for reading, called from sub Search::Elasticsearch::Bulk::reindex at /home/metacpan/metacpan-api/bin/../lib/MetaCPAN/Script/Mapping.pm line 143. With vars: {'request' => {'body' => 'c2NhbjszOzY2NzQxNjc6V09Ta3pra1RUY2VUUG5qOW4wMkQtdzs2NjcyMTEyOlpscHdDb0NkUU5xRlVnbEk1Y01UdFE7NjY3NDE2ODpXT1NremtrVFRjZVRQbmo5bjAyRC13OzE7dG90YWxfaGl0czoxOTc1MjI3ODs=','serialize' => 'std','method' => 'GET','ignore' => [],'mime_type' => 'application/json','path' => '/_search/scroll','qs' => {'scroll' => '1m'}},'status_code' => 599}
```

As far as I can see, the `reindex` doesn't take any timeout parameter for the `scroll_helper` it creates, but I'm not sure this is the problem.
Is there any workaround for it? (will try patching locally and retry).

Best,
Mickey

**Comment by clintongormley** [comment]
Hi @mickeyn 

This is on a client where everything else is working?  It looks like retrieving the results from a scroll request is taking a long time (more than 30s by default I think?)

What version of ES are you on?  And which Cxn backend are you using?

You can try setting [`request_timeout`](https://metacpan.org/pod/Search::Elasticsearch::Role::Cxn#request_timeout)  to a higher value when you create the client, but it'd be good to figure out why it is taking more than 30s.  Perhaps the docs are too big and you should just reduce the size of the scroll?  (In `source` set `size` to something lower)

**Comment by mickeyn** [comment]
`ES 2.4.0`
AFAIK, the backend used is `HTTPTiny`.
The docs may be too big, so if it fails again, I will try those suggested settings as well.
I'll let you know how it goes,
Thanks.

**Comment by mickeyn** [comment]
so far tried *unsuccessfully*:

`request_timeout => 300` (object creation)
`size => 25` (source)

That's on-top of the `scroll => '30m'` which is locally patched to the `Bulk` module (the scroller creation in `reindex`)

Trying with more aggressive numbers.

Note: since the last change, I no longer get the timeout message, but a long (too long to see the start of) output of the documents, I suspect it's the `size`, so now trying with 5 and logging into a file.

**Comment by clintongormley** [comment]
Btw the size is per shard. If you sort by _doc then it'll be a global size

**Comment by mickeyn** [comment]
still timing out.
got as low as `size => 5` and `request_timeout => 600`.

**Comment by mickeyn** [comment]
well, nothing seems to beat this one. the data in this type that consistently fail (after a couple of hours of running) is big, but i'm not sure it's big enough to be a challenge for reindexing.

what are my other options?
- should I move the data by scrolling one index and creating it into the other myself?
- is there a workaround to deleting a `type` mapping in the index and recreating it? (this is where the whole problem started, we want to change mapping of a type which is not as big as the one failing to reindex, and I can recreate its data, so I don't care about losing this type with all its content)

**Comment by clintongormley** [comment]
@mickeyn are you seeing anything in the elasticsearch logs?  This doesn't sound client related.  If you try the same thing with curl (you can see the curl commands if you set `trace_to => Stderr`) you should see similar issues (if you don't then we need to investigate more).

> should I move the data by scrolling one index and creating it into the other myself? (without using the reindex method)

this is essentially what reindex is doing - it does a scrolled search and then a bulk index.  Hence asking about what you see in the ES logs. I'm wondering if you're running out of memory or having slow GCs.

> is there a workaround to deleting a type mapping in the index and recreating it?

You can't delete types anymore and you can't change existing mappings, but you can delete all of the documents of that type.  Short of that, reindexing is your only hope.  Oh you're on 2.4, what about trying the built in reindex command (instead of the Perl version)?

**Comment by mickeyn** [comment]
OK, I think I found my solution:

I'm doing my own scroll and create (copying `id` and `source`) and doing it in batches according to https://www.elastic.co/guide/en/elasticsearch/guide/current/reindex.html

The one problematic type is now cloned, after batching it yearly and sometimes monthly. the copy of the id allowed my to tolerate a few timeouts (that may have been caused by load on the cluster) and repeat the failing batches without worrying about data duplication (couldn't find a 'skip' flag I can use for suppressing failure messages for `already exists`).

For now, that will do. I still need to run some validation, but by the doc numbers it seems good. 

Thanks,
Mickey
